blob/service.go

```diff
@@ -0,0 +1,279 @@
+)
+
+var (
+	ErrBlobNotFound = errors.New("blob: not found")
+
+	log = logging.Logger("blob")
+)
+
+// TODO(@vgonkivs): remove after bumping celestia-app
+// defaultMinGasPrice is the default min gas price.
+const defaultMinGasPrice = 0.001
+
+// headerGetter fetches header by the provided height
+type headerGetter func(context.Context, uint64) (*header.ExtendedHeader, error)
+
+type Service struct {
+	// accessor dials the given celestia-core endpoint to submit blobs.
+	accessor *state.CoreAccessor
+	// shareGetter retrieves the EDS to fetch all shares from the requested header.
+	shareGetter  share.Getter
+	headerGetter headerGetter
+}
+
+func NewService(
+	state *state.CoreAccessor,
+	getter share.Getter,
+	headGetter headerGetter,
+) *Service {
+	return &Service{
+		accessor:     state,
+		shareGetter:  getter,
+		headerGetter: headGetter,
+	}
+}
+
+// Submit sends PFB transaction and reports the height in which it was included.
+// Allows sending multiple Blobs atomically synchronously.
+// Uses default wallet registered on the Node.
+func (s *Service) Submit(ctx context.Context, blobs ...*Blob) (uint64, error) {
+	log.Debugw("submitting blobs", "amount", len(blobs))
+
+	var (
+		gasLimit = estimateGas(blobs...)
+		fee      = int64(defaultMinGasPrice * float64(gasLimit))
+		b        = make([]*apptypes.Blob, len(blobs))
+	)
+
+	for i, blob := range blobs {
+		b[i] = &blob.Blob
+	}
+
+	resp, err := s.accessor.SubmitPayForBlob(ctx, types.NewInt(fee), gasLimit, b...)
+	if err != nil {
+		return 0, err
+	}
+	return uint64(resp.Height), nil
+}
+
+// Get retrieves all the blobs for given namespaces at the given height by commitment.
+func (s *Service) Get(ctx context.Context, height uint64, nID namespace.ID, commitment Commitment) (*Blob, error) {
+	blob, _, err := s.getByCommitment(ctx, height, nID, commitment)
+	if err != nil {
+		return nil, err
+	}
+	return blob, nil
+}
+
+// GetProof retrieves all blobs in the given namespaces at the given height by commitment
+// and returns their Proof.
+func (s *Service) GetProof(
+	ctx context.Context,
+	height uint64,
+	nID namespace.ID,
+	commitment Commitment,
+) (*Proof, error) {
+	_, proof, err := s.getByCommitment(ctx, height, nID, commitment)
+	if err != nil {
+		return nil, err
+	}
+	return proof, nil
+}
+
+// GetAll returns all blobs under the given namespaces at the given height.
+// GetAll can return blobs and an error in case if some requests failed.
+func (s *Service) GetAll(ctx context.Context, height uint64, nIDs ...namespace.ID) ([]*Blob, error) {
+	header, err := s.headerGetter(ctx, height)
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		blobs     = make([]*Blob, 0)
+		ch        = make(chan []*Blob)
+		resultErr = make([]error, 0, len(nIDs))
+		errCh     = make(chan error)
+	)
+
+	for _, nID := range nIDs {
+		go func(nID namespace.ID) {
+			log.Infow("requesting blobs", "height", height, "nID", nID.String())
+			err := s.getBlobs(ctx, nID, header.DAH, ch)
```

Comment:
i'd rather s.getBlobs returns a blob, err and then blob gets sent to blobCh in case of nil err instead of chan getting passed into the method.

Reply:
but that is how it was done previously

Reply:
reworked